<?xml version="1.0" encoding="UTF-8"?>
<!-- This spawns an MAV and runs all nodes associated with the MAV -->
<launch>
  <arg name="mav_name"            default="mav0"/>
  <arg name="model_name"          default="chiny"/>
  <arg name="enable_ground_truth" default="true" />
  <arg name="log_file"            default="$(arg mav_name)" />
  <arg name="color"               default="White"/>
  <arg name="x"                   default="0"/>
  <arg name="y"                   default="0"/>
  <arg name="z"                   default="0.2"/>
  <arg name="yaw"                 default="0"/>

  <group ns="$(arg mav_name)">
    <!-- Spawn in Gazebo -->
    <include file="$(find fcu_sim)/launch/spawn_mav.launch">
      <arg name="mav_name"            value="$(arg mav_name)" />
      <arg name="model"               value="$(find desktopquad_sim)/agents/$(arg model_name)/$(arg model_name).xacro" />
      <arg name="x"                   value="$(arg x)"/>
      <arg name="y"                   value="$(arg y)"/>
      <arg name="z"                   value="$(arg z)"/>
      <arg name="yaw"                 value="$(arg yaw)"/>
      <arg name="color"               value="$(arg color)"/>
    </include>

    <!-- Load common parameters (mass, thrust, rotor configuration) -->
    <rosparam command="load" file="$(find desktopquad_sim)/agents/$(arg model_name)/$(arg model_name).yaml"/>

    <!-- Is Flying Publisher -->
    <node pkg="rostopic" type="rostopic" name="is_flying_pub" args="pub is_flying std_msgs/Bool true"/>

    <!-- Estimator -->
    <!-- <rosparam command="load" file="$(find ros_copter)/params/mekf.yaml" ns="estimator"/>
    <node name="estimator" pkg="ros_copter" type="mekf" output="screen">
      <param name="euler_integration_steps" value="20"/>
    </node>-->

    <!-- ArUco Localization -->
    <node name="aruco" pkg="aruco_localization" type="aruco_localization" output="screen">
      <param name="show_output_video" value="true" />
      <param name="markermap_config" value="$(find desktopquad_sim)/params/aruco_mip_36h12_markermap.yaml" />
      <param name="marker_size" value="0.0298" />

      <remap from="input_image" to="camera/rgb" />
    </node>

    <!-- PID Position Controller -->
    <rosparam command="load" file="$(find desktopquad_sim)/agents/$(arg model_name)/$(arg model_name)_controller.yaml" ns="controller"/>
    <node name="controller" pkg="desktopquad" type="controller" output="screen">
      <remap from="estimate" to="ground_truth/odometry"/>
    </node>

    <!-- Waypoint Manager -->
    <node name="waypoint_manager" pkg="desktopquad" type="waypoint_manager.py" output="screen">
      <rosparam command="load" file="$(find desktopquad_sim)/params/waypoints.yaml"/>
      <remap from="waypoint" to="high_level_command"/>
      <remap from="state" to="ground_truth/odometry"/>
    </node>

  </group>

</launch>